cuML/src/tsvd/tsvd.h

```diff
@@ -142,116 +141,99 @@ void signFlip(math_t *input, int n_rows, int n_cols, math_t *components,
-		math_t *singular_vals, paramsTSVD prms, cublasHandle_t cublas_handle,
-                      cusolverDnHandle_t cusolver_handle) {
-    ///@todo: make this to be passed via the interface!
-    DeviceAllocator mgr = makeDefaultAllocator();
-
-        tsvdFit(input, components, singular_vals, prms, cublas_handle, cusolver_handle);
-	tsvdTransform(input, components, trans_input, prms, cublas_handle);
-
-	signFlip(trans_input, prms.n_rows, prms.n_components, components,
-			prms.n_cols);
-
-	math_t *mu_trans;
-	allocate(mu_trans, prms.n_components);
-
-	Stats::mean(mu_trans, trans_input, prms.n_components, prms.n_rows, true,
-			false);
-	Stats::vars(explained_var, trans_input, mu_trans, prms.n_components,
-			prms.n_rows, true, false);
-
-	math_t *mu;
-	allocate(mu, prms.n_cols);
-	math_t *vars;
-	allocate(vars, prms.n_cols);
-
-	Stats::mean(mu, input, prms.n_cols, prms.n_rows, true, false);
-	Stats::vars(vars, input, mu, prms.n_cols, prms.n_rows, true, false);
-
-	math_t *total_vars;
-	allocate(total_vars, 1);
-	Stats::sum(total_vars, vars, 1, prms.n_cols, false);
-
-	math_t total_vars_h;
-	updateHost(&total_vars_h, total_vars, 1);
-	math_t scalar = math_t(1) / total_vars_h;
-
-	LinAlg::scalarMultiply(explained_var_ratio, explained_var, scalar,
-			prms.n_components);
-
-	CUDA_CHECK(cudaFree(mu_trans));
-	CUDA_CHECK(cudaFree(mu));
-	CUDA_CHECK(cudaFree(vars));
-	CUDA_CHECK(cudaFree(total_vars));
+void tsvdFitTransform(const cumlHandle_impl& handle, math_t *input,
+                      math_t *trans_input, math_t *components,
+                      math_t *explained_var, math_t *explained_var_ratio,
+                      math_t *singular_vals, paramsTSVD prms) {
+    tsvdFit(handle, input, components, singular_vals, prms);
+    tsvdTransform(input, components, trans_input, prms, handle.getCublasHandle());
+
+    signFlip(trans_input, prms.n_rows, prms.n_components, components,
+             prms.n_cols);
+
+    device_buffer<math_t> mu_trans(handle.getDeviceAllocator(), handle.getStream(),
+                                   prms.n_components);
+    Stats::mean(mu_trans.data(), trans_input, prms.n_components, prms.n_rows, true,
+                false);
+    Stats::vars(explained_var, trans_input, mu_trans.data(), prms.n_components,
+                prms.n_rows, true, false);
+
+    device_buffer<math_t> mu(handle.getDeviceAllocator(), handle.getStream(),
+                             prms.n_cols);
+    device_buffer<math_t> vars(handle.getDeviceAllocator(), handle.getStream(),
+                               prms.n_cols);
+
+    Stats::mean(mu.data(), input, prms.n_cols, prms.n_rows, true, false);
+    Stats::vars(vars.data(), input, mu.data(), prms.n_cols, prms.n_rows, true, false);
+
+    device_buffer<math_t> total_vars(handle.getDeviceAllocator(), handle.getStream(), 1);
+    Stats::sum(total_vars.data(), vars.data(), 1, prms.n_cols, false);
+
+    math_t total_vars_h;
+    updateHost(&total_vars_h, total_vars.data(), 1);
```

Comment:
synchronous `updateHost` does not take a stream. This needs to be

```
updateHostAsync( ..., stream);
cudaStreamSynchronize(stream);
```
if I do not miss anything.